packages/eslint-plugin/src/rules/no-invalid-void-type.ts

```diff
@@ -61,6 +61,40 @@ export default util.createRule<[Options], MessageIds>({
       AST_NODE_TYPES.Identifier,
     ];
 
+    /*
+     * isValidUnionType: checks that each member of the union is either a void
+     * or a Promise<void>
+     */
+
+    function isValidUnionType(node: TSESTree.TSUnionType): boolean {
+      function isValidMemberType(member: TSESTree.TypeNode): boolean {
+        if (member.type === AST_NODE_TYPES.TSVoidKeyword) {
+          return true;
+        }
+
+        if (
+          member.type === AST_NODE_TYPES.TSTypeReference &&
+          member.typeParameters?.type ===
+            AST_NODE_TYPES.TSTypeParameterInstantiation
+        ) {
+          const sourceCode = context.getSourceCode();
+          const fullyQualifiedName = sourceCode
+            .getText(member.typeName)
+            .replace(/ /gu, '');
+
+          return (
+            fullyQualifiedName === 'Promise' &&
+            member.typeParameters.params.length === 1 &&
+            member.typeParameters.params[0].type ===
+              AST_NODE_TYPES.TSVoidKeyword
+          );
+        }
+        return false;
+      }
```

Comment:
Hmmm.
I think this should follow the same rule as the base check.
i.e. instead of special casing `Promise`, it should allow all generic types if `allowInGenericTypeArguments: true`, or just the whitelisted ones if `allowInGenericTypeArguments: string[]`.